File: Serene/Serene.Script/Northwind/Customer/CustomerDialog.cs

namespace Serene.Northwind
{
    using jQueryApi;
    using Serenity;
    using System.Collections.Generic;

    [IdProperty("ID"), NameProperty("CustomerID"), Flexify, Maximizable]
    [FormKey("Northwind.Customer"), LocalTextPrefix("Northwind.Customer"), Service("Northwind/Customer")]
    public class CustomerDialog : EntityDialog<CustomerRow>
    {
        private string loadedState;
        private CustomerOrdersGrid ordersGrid;

        public CustomerDialog()
        {
            ordersGrid = new CustomerOrdersGrid(this.ById("OrdersGrid"));
            ordersGrid.Element.FlexHeightOnly();

            this.ById("NoteList").Closest(".field").Hide().End().AppendTo(this.ById("TabNotes"));

            DialogUtils.PendingChangesConfirmation(this.element, () => GetSaveState() != loadedState);

            tabs.OnActivate += (e, i) => this.Arrange();
        }

        private string GetSaveState()
        {
            try { return Q.ToJSON(GetSaveEntity()); } catch { return null; }
        }

        public override void LoadResponse(RetrieveResponse<CustomerRow> data)
        {
            base.LoadResponse(data);
            loadedState = GetSaveState();
        }

        protected override void LoadEntity(CustomerRow entity)
        {
            base.LoadEntity(entity);

            tabs.SetDisabled("Orders", this.IsNewOrDeleted);

            ordersGrid.CustomerID = entity.CustomerID;
        }

        protected override void OnSaveSuccess(ServiceResponse response)
        {
            base.OnSaveSuccess(response);

            Q.ReloadLookup(CustomerRow.LookupKey);
        }
    }
}

File: Serene/Serene.Web/Modules/Northwind/Customer/CustomerDialog.ts
namespace Serene.Northwind {

    @Serenity.Decorators.registerClass()
    export class CustomerDialog extends Serenity.EntityDialog<CustomerRow, any> {
        protected getFormKey() { return CustomerForm.formKey; }
        protected getIdProperty() { return CustomerRow.idProperty; }
        protected getLocalTextPrefix() { return CustomerRow.localTextPrefix; }
        protected getNameProperty() { return CustomerRow.nameProperty; }
        protected getService() { return CustomerService.baseUrl; }

        protected form = new CustomerForm(this.idPrefix);

        private ordersGrid: CustomerOrdersGrid;
        private loadedState: string;

        constructor() {
            super();

            this.ordersGrid = new CustomerOrdersGrid(this.byId('OrdersGrid'));
            this.ordersGrid.element.flexHeightOnly(1);

            this.byId('NoteList').closest('.field').hide().end().appendTo(this.byId('TabNotes'));
            DialogUtils.pendingChangesConfirmation(this.element, () => this.getSaveState() != this.loadedState);
        }

        getSaveState() {
            try {
                return $.toJSON(this.getSaveEntity());
            }
            catch (e) {
                return null;
            }
        }

        loadResponse(data) {
            super.loadResponse(data);
            this.loadedState = this.getSaveState();
        }

        loadEntity(entity: CustomerRow) {
            super.loadEntity(entity);

            Serenity.TabsExtensions.setDisabled(this.tabs, 'Orders', this.isNewOrDeleted());

            this.ordersGrid.customerID = entity.CustomerID;
        }

        onSaveSuccess(response) {
            super.onSaveSuccess(response);

            Q.reloadLookup('Northwind.Customer');
        }
    }
}